Update local file and url download

Jodasue opened the pull request:
1. Updated the local file and made sure it was supported int TazerFile and lib.
2. Added support in cmake for CURL and SSL.  We must have SSL support to handle https files.  For now, we require both CURL and SSL.  If we find that we only care about http, then we could just use CURL.
3. Fixed some bugs in the python run script.  This script could still use some work, but is a strong start.

Commit:
1. Updated to use LocalFiles
2. Added cmake support to use CURL and SSL
3. Fixed minor bugs in python scripts

Commit:
Merge branch 'master' of https://github.com/pnnl/tazer into UpdateLocalFileAndURLDownload

Commit:
1. Adding a UrlFileCache which inherits form LocalFileCache
2. Modified LocalFile to use UrlFileCache when CURL is available
3. Fixed deleting TazerFile -- Needed to use name to lookup not metaName

Jodasue commented:
Added a second commit with a UrlDownload cache.  This cache downloads the whole file first, and then acts like a normal LocalFileCache.  This is a first approach... Next version we should not store the file in the file system, rather let the blocks live in the cache.

Broader comment, on if we want the downloading capability to be a cache?  Do we like the overloading of a regular cache method?  This means we have to change the code in the InputFile - to match LocalFile when adding caches.

Commit:
Fixing some printf in URL files

rdfriese commented on src/client/Lib.cpp:
```diff
@@ -150,6 +151,7 @@ void __attribute__((constructor)) tazerInit(void) {
         timer.start();
         Loggable::mtx_cout = new std::mutex();
         InputFile::_cache = new Cache(BASECACHENAME, CacheType::base);
+        LocalFile::_cache = new Cache(BASECACHENAME, CacheType::base);
```
We might want to think about the implications of creating two separate caches...
do we want to allow both input and local at the same time?
thinking about it, can local just collapse down to inputfile just using "localfilecache/urlcache" as the llc?

rdfriese commented on src/client/LocalFile.cpp:
```diff
@@ -108,50 +111,55 @@
         LocalFile::_cache->addCacheLevel(c, ++level);
     }
 
+    #ifdef USE_CURL
```
presumably we make this Config variable? (or this is set by CMAKE?

rdfriese commented:
I agree with your first sentence, and think that we want to support both modes.

for the second sentence I think i might have asked something similar in one of the source comments I left. It might be reasonable to remove LocalFile altogether, and just fold that functionality into InputFile, and then just have the LocalFileCache or URLCache as the last level

nathantallent commented:
As a general semantic comment, I do not think the download should be a cache: 1. Normal caches are bounded and are not guaranteed to hold an entire file. 2. Caches are normally used by multiple threads, so we would have to "pin" the blocks to ensure earlier parts of the file are not evicted while later parts are still downloading.

I could imagine reusing the implementation (if convenient) of an unbounded cache, which is perhaps what you were thinking.

> Added a second commit with a UrlDownload cache. This cache downloads the whole file first, and then acts like a normal LocalFileCache. This is a first approach... Next version we should not store the file in the file system, rather let the blocks live in the cache.
> 
> Broader comment, on if we want the downloading capability to be a cache? Do we like the overloading of a regular cache method? This means we have to change the code in the InputFile - to match LocalFile when adding caches.

Commit:
1. Adding UrlCache to handle ranges
2. Added some test for range copy
3. Added curl options to stay open for program instead of closing
4. Added support for downloading to the server

Jodasue committed:
Merge branch 'master' into UpdateLocalFileAndURLDownload

Commit:
Fixing the non-curl compile path.

Commit:
Merge branch 'master' of https://github.com/pnnl/tazer into UpdateLocalFileAndURLDownload

Commit:
Merge branch 'UpdateLocalFileAndURLDownload' of https://github.com/pnnl/tazer into UpdateLocalFileAndURLDownload
